proposals/cli_schema.md

```diff
@@ -0,0 +1,248 @@
+
+None
+
+## Motivation
+
+[motivation]: #motivation
+
+Test manifests today have no formalized structure and often require multiple files to declare expected test results (ex., `kyverno-test.yaml`, `values.yaml`, `user_info.yaml`, etc). Because there's no formalized structure, writing test cases is difficult because there's no linting or validation of the contents. This creates a poor user experience. Additionally, when adding new capabilities, it is currently not possible to version that structure which creates breaking changes for everyone. These changes will support anyone who wishes to write test cases for the `kyverno test` command. The outcome expected is a formalized, versioned schema with full validation similar to how Kubernetes and other systems work today.
+
+## Proposal
+
+This provides a high level overview of the feature.
+
+A new schema is being proposed for the `test` command which will provide the necessary structure to account for all major current capabilities of Kyverno and close some issues related to problems. There are some high-level functional requirements that must be met.
+
+1. The structure must be completely validated.
+2. It must allow backwards compatibility with existing test manifests out in the wild.
+   1. A test manifest which supplies the `apiVersion` and `kind` as specified in the proposed schema below is not subject to this backwards compatibility; only those in which these fields are absent should be subject to the existing logic.
+3. All file names must be checked for existence and validity.
+4. Objects in the `results[]` section should be validated against the rule type to determine if the necessary mandatory fields are present.
+5. Policies which are of type `Policy` must specify Namespace.
+6. There must not be duplicate names in resources.
+7. Deprecated fields must be handled with warnings to users.
+8. Breaking changes should cause the apiVersion to be incremented.
+
+```yaml
+apiVersion: cli.kyverno.io/v1beta1
+kind: KyvernoTest
+metadata:
+  name: test-check-nvidia-gpus
+  labels:
+    foolabel: foovalue
+  annotations:
+    fookey: foovalue
+spec:
+  policies:
+    # support fully-qualified file paths (https://github.com/kyverno/kyverno/issues/2315)
+    # wherever files are the value. Should be relative and absolute.
+  - check-nvidia-gpu.yaml
+  - relative/path/to/next/policy.yaml
+  - /fully/qualified/path/to/next/policy.yaml
+  resources:
+  - resource.yaml
+  - relative/path/to/next/resource.yaml
+  - /fully/qualified/path/to/next/resource.yaml
+  results:
+  # results[] vary depending on the rule type being tested.
+  # validate
+  - policy: policyname01 # required
+    rule: validaterulename # required
+    # resources must support multiple resources, all of which
+    # should be subject to the rest of the field values 
+    # in this result. (https://github.com/kyverno/kyverno/issues/2857)
+    resources: # required
+    - resource01
+    - resource02
+    - resource03
```

Comment:
I would like to see something like - 

```
resources:
   # If just object is present, mode is create
   # if both object and old is present, mode is update
   # If only old is present, mode is delete
   - object: <resource-pool>:apiversion/group/<namespace>/<name>
     old: <resource-pool>:apiversion/group/<namespace>/<name>
     patched: <resource-pool>:apiversion/group/<namespace>/<name>
     cloneSource: <resource-pool>:apiversion/group/<namespace>/<name>
     generated: <resource-pool>:apiversion/group/<namespace>/<name>
```

where resource-pool is an optional identifier that can be used to group resources.

Reply:
This appears related to the [other](https://github.com/kyverno/KDP/pull/21#discussion_r892628405) comment. Not sure I get the resource pool concept or the object item there.

Reply:
Bump @samj1912

Reply:
Tried to capture this in https://github.com/kyverno/KDP/pull/21/commits/24a2d5464ceaf22f8e030228d5c4d7518055bc53

Reply:
Resolving.